### PR TITLE
Attempt to fix truncated Jedi LSP responses

### PR DESCRIPTION
**Pull request by dpriedel**
## References

#450

## Code changes

This patch attempts to fix the truncated Jedi LSP responses which
produce warning messages about content-length mismatch. It does so
by looping around and trying again upon an empty read. The theory
is that the content producer is just not ready with more data yet.

- add function parameter 'max_empty_parts_in_a_row' with a default
value of 50.  This will be decremented on each empty read.
- change the default value of the 'max_parts' parameter to 100
since we should get larger chunks on each successful read.

Either of these values could be adjusted based on experience.

### CODE REVIEW

File: python_packages/jupyter_lsp/jupyter_lsp/stdio.py [hunk 1]
```diff
@@ -116,10 +116,11 @@ def _read_content(self, length: int, max_parts=1000) -> Optional[bytes]:
         raw = None
         raw_parts: List[bytes] = []
         received_size = 0
-        while received_size < length and len(raw_parts) < max_parts:
+        while received_size < length and len(raw_parts) < max_parts and max_empty_parts_in_a_row > 0:
             part = self.stream.read(length)
             if part is None:
-                break  # pragma: no cover
+                --max_empty_parts_in_a_row
```
Comment (krassowski) [inline python_packages/jupyter_lsp/jupyter_lsp/stdio.py:122]: `-= 1`? `--` gives + in python

**Comment by bollwyvl** [comment]
Thanks for the PR :heart:! This has been a bear, and we really appreciate the help.

Aside from the text thing above (please try `python -m scripts.lint from the repo`, see what it says), we definitely need to put this under a representative test, otherwise it's hard to keep it fixed.

We don't have a concrete example in the repo yet, but we should figure out a test that:

- spins up _at least two_ language servers, so we don't fool ourselves on the blocking aspects, especially on windows
  - they _can_ be the same implementation, just different spec names, i guess
- repeatedly generates a big enough response to demonstrate this condition 

@krassowski any thoughts? I'm most comfortable trying to test vs pyls, but I know you had some good gotchas with R...

**Comment by krassowski** [comment]
I was thinking about just copying the large response from R server that I can locally reproduce. An example from @dpriedel of an action that triggers the problem with pyls would be very helpful too.

I am not sure if the buffer length always stays constant (DEFAULT_BUFFER_SIZE), this would need to be accounted for when designing a unit tests; we could simulate a delay using some sleep in a thread.

On a more general note, and as mentioned elsewhere the `BufferedIOBase` seems to be troublesome (https://bugs.python.org/issue13322) so we may need to:
 - experiment with `read1` instead of `read`
 - experiment with row (non-buffered) stream (`RawIOBase`) <-- I have hopes that this might have a positive performance impact but I might be wrong
 - use another async communication implementation

so the priority would be:
1) a test for large message on this PR
2) maybe a test handling buffered IO delays as addressed by the issue in this PR, but may need rewriting if we decide against using buffered IO in the future

**Comment by bollwyvl** [comment]
Ha, my favorite stress test is `astropy.units.<TAB>` https://docs.astropy.org/en/stable/units/#module-astropy.units

File: python_packages/jupyter_lsp/jupyter_lsp/stdio.py (hunk 1 above)
Comment (dpriedel) [reply python_packages/jupyter_lsp/jupyter_lsp/stdio.py:122]: I don't understand your comment


python
Python 3.9.1 (default, Dec 13 2020, 11:55:53)
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> counter = 10
>>> while counter > 0:
...  print(counter)
...  counter -= 1
...
10
9
8
7
6
5
4
3
2
1
>>> print(counter)
0
>>>

Comment (krassowski) [reply python_packages/jupyter_lsp/jupyter_lsp/stdio.py:122]: Yes, `x -= 1` is the right way to do this, but you seem to have accidentally used `--x` in your commit (which is a common muscle memory for people who program in Python and another C-like language)

**Comment by dpriedel** [comment]
gah !!


sorry about that

**Comment by dpriedel** [comment]
> Aside from the text thing above (please try `python -m scripts.lint from the repo`, see what it says),
> 


I tried to run that script but it seems to require several modules (robot...) that I don't have on my system.

**Comment by dpriedel** [comment]
> Aside from the text thing above (please try `python -m scripts.lint from the repo`, see what it says)

I think the code change is clean now -- no more flake8 warnings.

I got the lint script to run on my system. Attached is the output showing zero errors on my system

I hope this is useful.

[lsp.log](https://github.com/krassowski/jupyterlab-lsp/files/5875541/lsp.log)

**Comment by krassowski** [comment]
I got a nice reproduction on Windows run:

```
>       assert result == message * repeats
E       AssertionError: assert 'abababababababababababababababababababababababababababab' == 'ababababababababababababababababababababababababababababababababababababababababababababab'
E         - ababababababababababababababababababababababababababababababababababababababababababababab
E         ?                                                         ----------------------------------
E         + abababababababababababababababababababababababababababab

python_packages\jupyter_lsp\jupyter_lsp\tests\test_stdio.py:81: AssertionError
------------------------------ Captured log call ------------------------------
WARNING  traitlets:stdio.py:141 Readout and content-length mismatch: 58 vs 90
```

The problem is that it still happens. Another problem is that the tests have timeout issues.

File: python_packages/jupyter_lsp/jupyter_lsp/stdio.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ class LspStdIoReader(LspStdIoBase):
 
     @default("max_wait")
     def _default_max_wait(self):
-        return 2.0 if os.name == "nt" else self.min_wait
+        return 0.1 if os.name == "nt" else self.min_wait * 2
```
Comment (krassowski) [inline python_packages/jupyter_lsp/jupyter_lsp/stdio.py:64]: Ok, so this is ridiculous right now (on purpose). @bollwyvl do you remember why Windows was set to `2.0`? I think this is not needed anymore. I think that we could just go with `0.2` or rather `self.min_wait * 4`?

**Comment by krassowski** [comment]
Thank you @dpriedel! After adding tests and disabling buffering all now runs smoothly across all the operating systems.

**Pull request merged**